subsys/task_wdt/task_wdt.c

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright (c) 2020 Libre Solar Technologies GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "task_wdt/task_wdt.h"
+
+#include <drivers/watchdog.h>
+#include <power/reboot.h>
+#include <device.h>
+#include <errno.h>
+
+#define LOG_LEVEL CONFIG_WDT_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(task_wdt);
+
+/*
+ * This dummy channel is used to continue feeding the hardware watchdog if the
+ * task watchdog timeouts are too long for regular updates
+ */
+#define TASK_WDT_BACKGROUND_CHANNEL (-1)
+
+/*
+ * Task watchdog channel data
+ */
+struct task_wdt_channel {
+	/* period in milliseconds used to reset the timeout, set to 0 to
+	 * indicate that the channel is available
+	 */
+	uint32_t reload_period;
+	/* abs. ticks when this channel expires (updated by task_wdt_feed) */
+	int64_t timeout_abs_ticks;
+	/* user data passed to the callback function */
+	void *user_data;
+	/* function to be called when watchdog timer expired */
+	task_wdt_callback_t callback;
+};
+
+/* array of all task watchdog channels */
+static struct task_wdt_channel channels[CONFIG_TASK_WDT_CHANNELS];
+
+/* timer used for watchdog handling */
+static struct k_timer timer;
+
+#ifdef CONFIG_TASK_WDT_HW_FALLBACK
+/* pointer to the hardware watchdog used as a fallback */
+static const struct device *hw_wdt_dev;
+static int hw_wdt_channel;
+static bool hw_wdt_started;
+#endif
+
+/**
+ * @brief Task watchdog timer callback.
+ *
+ * If the device operates as intended, this function will never be called,
+ * as the timer is continuously restarted with the next due timeout in the
+ * task_wdt_feed() function.
+ *
+ * If all task watchdogs have longer timeouts than the hardware watchdog,
+ * this function is called regularly (via the background channel). This
+ * should be avoided by setting CONFIG_TASK_WDT_MIN_TIMEOUT to the minimum
+ * task watchdog timeout used in the application.
+ *
+ * @param timer_id Pointer to the timer which called the function
+ */
+static void task_wdt_trigger(struct k_timer *timer_id)
```

Comment:
As far as I understand, the `k_timer` expiry functions are called from a system clock ISR context, which is already highest priority.
Using a peripheral timer could still be a nice option to reduce computing effort. However, I'm not sure how well timer peripherals are supported for each platform in Zephyr. The use of `k_timer` makes it very generic.

Reply:
Yeah, I like the use of k_timer by default for it's portability too. I could see being able to override with a peripheral timer when available  which would allow one to even catch hangs in timer callbacks or ISRs at a priority if configured correctly.